### PR TITLE
Integrate cognito

### LINKED ISSUE #13: Integrate cognito functionality into django user model
### Overview
We want to implement the user model in a way that allows the backend to use cognito authentication for users.

### Action Items
- [x] Follow the django cognito guide in the resources, using cognito token as username
- [ ] Maybe see if it can work with the VRMS client that talks to cognito
- [x] cleanup and documentation

### Resources/Instructions
- See the interaction diagram [here](https://djangostars.com/blog/bootstrap-django-app-with-cognito/). The frontend doesn't login to the backend. The frontend logs in to cognito, gets tokens, and uses a token to communicate with the backend. The backend has a separate connection with cognito to authenticate users.
- [VRMS cognito issue](https://github.com/hackforla/VRMS/issues/636)

### PR DESCRIPTION
**Pull request by fyliu**
- Closes #13

Added the necessary fields in the custom user model to hold cognito information and the necessary code to authenticate valid user tokens.
When an authenticated user requests a restricted resource, django verifies the token against the JWKS that's downloaded on runserver and a stub user is created if the token is authenticated.

To test this, it's a tedious process with several steps
1. Run the backend by calling ./scripts/buildrun.sh
1. Create a user in cognito user pool [here](https://hackforla-vrms-dev.auth.us-west-2.amazoncognito.com/login?client_id=3e3bi1ct2ks9rcktrde8v60v3u&response_type=token&scope=openid&redirect_uri=http://localhost:8000/admin).
    1. The required fields are: email as username, and a temporary password
1. Login and be redirected to localhost:8000....
1. Copy the URL into a text editor and copy the access_token, or paste the URL into [this tool](https://regexr.com/6ro69) that extracts the access_token.
1. Run this curl command (note the ending slash character):

    curl -H "Authorization: Bearer <ACCESS_TOKEN>" http://localhost:8000/api/v1/me/

1. The request should be successful if the token is authenticated
1. A new user should have been created. Run this command to see it (there should be 2 row if the testadmin user was created):

    docker-compose exec db psql --username=people_depot --dbname=people_depot_dev -c "select username from core_user"
1. See also the wiki page on [cognito authentication pre-deployment](https://github.com/hackforla/peopledepot/wiki/Cognito-authentication-workflow-(pre-deployment)).

### CODE REVIEW

**Comment by JasonEb** [comment]
image
ran into issues while running `buildrun.sh`. However manually running the commands within worked

**Comment by fyliu** [comment]
It looks like an issue with our machines having different setup. I'm on linux and you're on wsl2. This could happen since the scripts run outside docker.

I see that the way we get SCRIPT_DIR is using a bash-specific variable BASH_SOURCE. Maybe the wsl2 bash is different?

Could you try to run it but change `"${BASH_SOURCE[0]}"` to `"$0"`. Maybe also add `echo $SCRIPT_DIR` after the line that defines SCRIPT_DIR. That would give you an output of what the SCRIPT_DIR variable contains at that point.

**Comment by Neecolaa** [comment]
Tested on MacOS 12 using the instructions from https://github.com/hackforla/peopledepot/wiki/Cognito-authentication-workflow-(pre-deployment) instead of the ones in this pull request.
Screen Shot 2022-08-20 at 11 39 28
(The testadmin user doesn't show up so I have 1 user showing, not 2, but the API matches the results I got in the terminal)

**Comment by fyliu** [comment]
Squashed cognito-related commits and reordered commits so that the scripts are available here.

File: docs/architecture/authentication.md [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+4. Backend verifies token and processes request
+5. User gets response from authenticated API
+
+## Current Dev Setup
+
+1. Created app client called "backend within the vrms-dev user pool, with ALLOW_ADMIN_USER_PASSWORD_AUTH enabled
+2. "Domain Name" is already created at [https://hackforla-vrms-dev.auth.us-west-2.amazoncognito.com]
+3. In "App client settings", enabled Implicit grant and openid, Callback URL [http://localhost:8000/admin]
+
+## How it works now with the dev user pool and local development backend
```
Comment (cnk) [inline docs/architecture/authentication.md:29]: Fantastic instructions! Worked first time for me.

**Review by cnk — APPROVED** [review]
Super PR. I only found one quibble - and that is with the REQUIRED_FIELDS which are never required. If you could clean this up, that would be great but I am not going to hold off approving this just for that one line.

File: app/core/models.py [hunk 2]
```diff
@@ -1,5 +1,75 @@
...
+        "staff status",
+        default=False,
+        help_text="Designates whether the user can log into this admin site.",
+    )
+
+    objects = UserManager()
+
+    USERNAME_FIELD = "username"
+    EMAIL_FIELD = "email"
+    REQUIRED_FIELDS = ["email"]  # used only on createsuperuser
```
[the remark this answers is not in the document]
Comment (fyliu) [reply app/core/models.py:71]: You're right. We don't need this. We don't want to require email here anyway.

**Pull request merged**